llvm/include/llvm/IR/InstrTypes.h

```diff
@@ -692,6 +692,20 @@ class CastInst : public UnaryInstruction {
   }
 };
 
+/// Instruction that can have a nneg flag (only zext).
+class NonNegInstruction : public CastInst {
```

Comment:
Personally not a fan of `NonNegInstruction` as the name. To me atleast, that seems to imply a lot more coverage than just `zext`.
Maybe a bit more specific `NonNegCastInstruction`?

Reply:
Maybe it would be best to just drop it entirely? Given that we have just the one instruction supporting the flag, it might be better to stick with `isa<ZExtInst>` over `isa<NonNegInstruction>`. There is some risk that this may require future refactoring if we add the flag on additional instructions -- but that might be better than doing an unnecessary over-generalization now.

(We might want to add nneg on GEPs, but in that case the semantics of the flag would be sufficiently different that they could not share any common handling anyway.)

Reply:
> Is this meant to mirror how we use `OverflowingBinaryOperator` now?

Yes. I guess the name is confusing in that it sounds like "an instruction with the nneg flag", while in reality this is "an instruction that may or may not have the nneg flag". A more accurate name, taking inspiration from PossiblyExactOperator, would be PossiblyNonNegInstruction. Or I guess PossiblyNonNegCastInst taking your suggestion into account.

Reply:
Agreed, I would go with `PossiblyNonNegInstruction`. Think with that name (since its so explicitly a query) the `CastInst` isn't so important.